### PR TITLE
feat: Suggest upcast for assignable Java types

### PR DESCRIPTION
**Pull request by paulbutcher**
Related to #4392

### CODE REVIEW

**Comment by paulbutcher** [comment]
(the tests should pass after #4520 has been merged)

File: main/src/ca/uwaterloo/flix/language/errors/TypeError.scala [hunk 1]
```diff
@@ -185,6 +185,32 @@ object TypeError {
...
 
+  case class PossibleUpcast(expected: Type, inferred: Type, loc: SourceLocation) extends TypeError {
+    def summary: String = s"Expected type '${formatWellKindedType(expected)}' but found type: '${formatWellKindedType(inferred)}'."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""
+         |>> Expected type: '${red(formatWellKindedType(expected))}' but found type: '${red(formatWellKindedType(inferred))}'.
+         |
+         |'${formatWellKindedType(expected)}' appears to be assignable from '${formatWellKindedType(inferred)}'.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/TypeError.scala:196]: This message should go after the code fragment.

File: main/src/ca/uwaterloo/flix/language/errors/TypeError.scala [hunk 2]
```diff
@@ -185,6 +185,32 @@ object TypeError {
     def explain(formatter: Formatter): Option[String] = None
   }
 
+  case class PossibleUpcast(expected: Type, inferred: Type, loc: SourceLocation) extends TypeError {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/TypeError.scala:188]: Add documentation comment.

File: main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala [hunk 3]
```diff
@@ -200,7 +200,13 @@ object Unification {
     */
   def expectTypeM(expected: Type, actual: Type, loc: SourceLocation)(implicit flix: Flix): InferMonad[Type] = {
     def handler(e: TypeError): TypeError = e match {
-      case _: TypeError.MismatchedTypes => TypeError.UnexpectedType(expected, actual, loc)
+      case _: TypeError.MismatchedTypes => 
+        (expected, actual) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:204]: When looking at type constructors, it is more idiomatic to do:

`expected.typeConstructor, actual.typeConstructor match { case (Some(TypeCon.Native...)), ...) => `

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:204]: The reason to use `.typeConstructor` is that it allows us more easily to change the structure of types.

**Comment by magnus-madsen** [comment]
Looks good, can merge once the conflict is resolved :)

**Pull request merged**